### PR TITLE
Feat: DeduplicationEventProcessor

### LINKED ISSUE #480: DeduplicationEventProcessor
In order to remove duplicated events, for example caused by [this](https://github.com/getsentry/sentry-dart/pull/473#discussion_r638492342) we need an event processor which removes duplicated events.

See https://github.com/getsentry/sentry-java/blob/b45944c93d53f68bc1e64f5ac4c9a5400b252d44/sentry/src/main/java/io/sentry/DuplicateEventDetectionEventProcessor.java for reference

Depends on https://github.com/getsentry/sentry-dart/issues/479

### PR DESCRIPTION
**Pull request by ueman**
## :bulb: Motivation and Context
Closes https://github.com/getsentry/sentry-dart/issues/480

## :green_heart: How did you test it?
new tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (ueman) [inline dart/lib/src/event_processor/deduplication_event_processor.dart]: As we don't have a weak map or similar I think having a max length is an acceptable solution.
I haven't yet really thought about what a good length might be and just choose 20 for now.

File: dart/lib/src/sentry_options.dart
Comment (ueman) [inline dart/lib/src/sentry_options.dart]: This doc comment needs to be finished :D

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (ueman) [inline dart/test/event_processor/deduplication_event_processor_test.dart]: I'm pretty sure I'm missing some test cases.
Which events need deduplication and which don't?
As far as I can tell, Events without an exception aren't deduplicated and only events with the same exception are deduplicated.

Comment (ueman) [reply dart/test/event_processor/deduplication_event_processor_test.dart]: Also maybe a test case for the max length of the map.

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [inline dart/lib/src/event_processor/deduplication_event_processor.dart]: should we be specific about key and type?

Comment (marandaneto) [inline dart/lib/src/event_processor/deduplication_event_processor.dart]: can probably be internal too

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [inline dart/lib/src/event_processor/deduplication_event_processor.dart]: not sure if I understand this part, I don't see when the exception gets added to the map

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (ueman) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: The basetype of Exceptions in Dart is Object (the key) because you can throw anything. The value is never used. This is still just a naive conversion of the Java and C# implementations.

Maybe using a [LinkedHashSet](https://api.dart.dev/stable/2.13.3/dart-collection/LinkedHashSet-class.html) is a better option.

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (ueman) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: It's line 50

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [inline dart/lib/src/event_processor/deduplication_event_processor.dart]: looks like we don't use the value at all, wondering if `Queue` or `Dequeue` would make more sense

**Comment by marandaneto** [comment]
@bruno-garcia please review this one, which is a different approach due to the lack of WeakRef in Dart

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (marandaneto) [reply dart/test/event_processor/deduplication_event_processor_test.dart]: the main use case that should be deduplicated is like:

```
  private void test() {
    try {
      final int test = 1/0;
    } catch (Exception e) {
      // `e` is captured
      Sentry.captureException(e);
      throw e;
    }
  }
  
  private void myTest() throws Exception {
    // `test` throws and UncaughtExceptionHandlerIntegration gets called capturing `e` again
    // as `e` was already captured in the method `test`, the 2nd capture should be deduplicated
    test();
  }
```

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (ueman) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: Yep, I wondered about that in the Java and C# implementation, too.

Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: do we have a thread-safe type? Dart is sort of single-threaded but when using multiple isolates, this could be racey

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: oh damn, need coffee, indeed.
If you see the `Scope#_breadcrumbs`, I've used a `Queue` which is a circular buffer, looks like this approach could be similar to that, we could even offer the fixed size `20` configurable too.

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: https://github.com/getsentry/sentry-java/issues/1533 lets see if there was a reason

Comment (ueman) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: I think it's just to make look ups fast. For maps and sets you can just do a 'contains' lookup. While you can do that for lists too, it's not O(1) but O(n) in the worst case. At least if I remember the big O notation correctly 😅

Comment (ueman) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: Nope, there's no such thread safe class. Errors should be propagated via `Ports` (see https://api.dart.dev/stable/2.13.3/dart-isolate/dart-isolate-library.html) back to the main isolate. See for example Flutters [`compute` methods source code](https://github.com/flutter/flutter/blob/d79295af24c3ed621c33713ecda14ad196fd9c31/packages/flutter/lib/src/foundation/_isolates_io.dart).

Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: nice, so we just need to choose the right data structure for such a case.

Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: yep, that's a good point, let's see if it's the case :D

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: lets make 10, deduplicating events its a matter of micro/milli seconds, so 20 is quite high already, lets do 10 for now

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: as we don't want to hold references, nor to leak objects, after discussing, what would make more sense is having a data structure that stores the `hashCode` as an Int, it's way cheaper than the reference of the object in memory.
let's make a circular buffer similar to Scope#breadcrumbs

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: see https://github.com/getsentry/sentry-dart/pull/498#discussion_r653314510

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (marandaneto) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: the reason was that Java and .NET don't have a List interface of weak references, but only a `WeakHashMap`, so when the GC is doing its job, it'd clean up the data automatically from the HashMap, also, because of lookup time.
to improve lookup time, we can use a data structure that is sorted or O(1), but instead of the object itself, we insert and check the hashCode of the object.
I believe Java could be also improved, see https://github.com/getsentry/sentry-java/issues/1533#issuecomment-863016739

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (bruno-garcia) [inline dart/lib/src/event_processor/deduplication_event_processor.dart]: Might be worthy of `info` level (warnings is a stretch, right?)

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (bruno-garcia) [inline dart/lib/src/event_processor/deduplication_event_processor.dart]: I'd suggest something like 5

The use case is when there's a:

```
Sentry.captureException(e);
throw e;
```
And the caller also does `captureException` so these usually happen one after the other.

**Review by bruno-garcia — COMMENTED** [review]
Differently then Java and .NET this approach doesn't rely on weakref (I understand it isn't an option here) but it means we'll be holding the reference of the N exceptions and hence 'leaking' all the object tree they hold for the duration of the process. Makes me wonder if this should be an opt-in.

**Comment by ueman** [comment]
> Differently then Java and .NET this approach doesn't rely on weakref (I understand it isn't an option here) but it means we'll be holding the reference of the N exceptions and hence 'leaking' all the object tree they hold for the duration of the process. Makes me wonder if this should be an opt-in.

@marandaneto already suggested to store the hashes of the exception instead of the exception itself. 

This is only needed for the Http Client, which is also opt in. So there's nothing which speaks against making this opt in.

File: dart/lib/src/event_processor/deduplication_event_processor.dart
Comment (ueman) [reply dart/lib/src/event_processor/deduplication_event_processor.dart]: I'll make it configureable and set the default value to 5.

File: dart/lib/src/sentry_options.dart
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart]:
```suggestion
  /// Whether [SentryEvent] deduplication is enabled.
```

File: dart/lib/src/sentry_options.dart
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart]: if 0 should be `enableDeduplication=false`, so I'd remove `=`?

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (marandaneto) [inline dart/test/event_processor/deduplication_event_processor_test.dart]: I'd not make `exceptionsToKeepForDeduplication` nullable, but its a test, your call :)

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (marandaneto) [inline dart/test/event_processor/deduplication_event_processor_test.dart]: if you need a new Fixture every time which is not the case, use a lateinit var and setup, otherwise a global Fixture that creates new Sut only

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (marandaneto) [inline dart/test/event_processor/deduplication_event_processor_test.dart]: goes to sentry_test.dart probably

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (marandaneto) [inline dart/test/event_processor/deduplication_event_processor_test.dart]: use a `setUp` or `tearDown`

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (marandaneto) [inline dart/test/event_processor/deduplication_event_processor_test.dart]: I dont get this comment

File: dart/lib/src/sentry_options.dart
Comment (ueman) [reply dart/lib/src/sentry_options.dart]: Good catch

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (ueman) [reply dart/test/event_processor/deduplication_event_processor_test.dart]: Why though? This is the only test in this file which needs it.

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (ueman) [reply dart/test/event_processor/deduplication_event_processor_test.dart]: The test doesn't work if `outerTestMethod` is passed as `appRunner` callback

File: dart/test/event_processor/deduplication_event_processor_test.dart
Comment (marandaneto) [reply dart/test/event_processor/deduplication_event_processor_test.dart]: not really, I see multiple tests using `Sentry.close()`
but I'd say this is less important when doing https://github.com/getsentry/sentry-dart/pull/498/files#r656023034

File: dart/lib/src/sentry_options.dart
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart]: tbh the name is a bit off for me, but I am not sure what to suggest either.
maybe @bruno-garcia or @philipphofmann ?

`maxDeduplicationQueueItems` looks better IMO

**Review by marandaneto — APPROVED** [review]
pre-approving, lets just figure out this before merging https://github.com/getsentry/sentry-dart/pull/498/files#r659610845

File: dart/lib/src/sentry_options.dart
Comment (ueman) [reply dart/lib/src/sentry_options.dart]: Maybe `maxDeduplicationItems`. It's not really relevant for the user how the exceptions are stored, isn't it?

Comment (philipphofmann) [reply dart/lib/src/sentry_options.dart]: `maxDeduplicationItems` also makes sense to me. Do we even need to make this option public?

Comment (ueman) [reply dart/lib/src/sentry_options.dart]: No, not necessarily. The value of it is just a guess for what should work well. So that's why I made it configureable.

**Comment by marandaneto** [comment]
sounds good, merge it :)

File: dart/lib/src/sentry_options.dart
Comment (marandaneto) [reply dart/lib/src/sentry_options.dart]: other languages have a WeakRef list so it's not necessary having such an option.
here the approach is different, and depending on how you use the SDK, 5 items may be enough, or not.

**Pull request merged**
